scst/src/scst_dlm.c

```diff
@@ -1494,6 +1615,199 @@ static void scst_pr_dlm_cleanup(struct scst_device *dev)
 	dev->pr_dlm = NULL;
 }
 
+static struct scst_dlm_rem_ua *scst_dlm_alloc_rem_ua(void)
+{
+	struct scst_dlm_rem_ua *ua = NULL;
+
+	ua = kzalloc(sizeof(struct scst_dlm_rem_ua),  GFP_KERNEL);
+	if (ua == NULL) {
+		PRINT_ERROR("%s", "Unable to allocate unit attention");
+		goto out;
+	}
+
+	TRACE_MEM("Allocated dlm ua %p", ua);
+
+	ua->lksb.lksb.sb_lvbptr = (void *)ua->lvb;
+	ua->lksb.lksb.sb_lkid = 0;
+
+out:
+	return ua;
+}
+
+static void scst_dlm_free_rem_ua(struct scst_dlm_rem_ua *ua)
+{
+	TRACE_MEM("Freeing dlm ua %p", ua);
+	kfree(ua);
+}
+
+/*
+ * scst_dlm_pr_reg_queue_rem_ua - queue a remote unit attention for transmition to the remote node
+ */
+static void scst_dlm_pr_reg_queue_rem_ua(struct scst_device *dev,
+					 struct scst_dev_registrant *reg,
+					 int key, int asc, int ascq)
+{
+	struct scst_dlm_rem_ua *ua = NULL;
+	struct scst_pr_dlm_data *const pr_dlm = dev->pr_dlm;
+
+	TRACE_ENTRY();
+
+	ua = scst_dlm_alloc_rem_ua();
+	if (ua == NULL) {
+		PRINT_ERROR("%s", "Unable to allocate unit attention for dlm registrant");
+		return;
+	}
+
+	ua->key = key;
+	ua->asc = asc;
+	ua->ascq = ascq;
+
+	/*
+	 * Lock to ensure we are not currently doing a scst_copy_to_dlm or
+	 * scst_copy_from_dlm.
+	 */
+	mutex_lock(&pr_dlm->ls_mutex);
+	list_add_tail(&ua->rem_ua_list_entry, &reg->pending_rem_ua_list);
+	mutex_unlock(&pr_dlm->ls_mutex);
+
+	TRACE_EXIT();
+}
+
+/*
+ * scst_dlm_pr_reg_release_rem_ua - remove from list and free remote unit attention data structure
+ */
+static void scst_dlm_pr_reg_release_rem_ua(struct scst_dlm_rem_ua *ua)
+{
+	TRACE_ENTRY();
+
+	TRACE_DBG("Releasing dlm ua %p", ua);
+	list_del(&ua->rem_ua_list_entry);
+	scst_dlm_free_rem_ua(ua);
+
+	TRACE_EXIT();
+}
+
+/*
+ * scst_dlm_pr_reg_send_rem_ua - send a unit attention to the registrant on a remote node
+ *
+ * This will create a lock containing the necessary UA information, increment next_rem_ua_idx and
+ * add the struct scst_dlm_rem_ua to the registrant's sent_rem_ua_list.
+ */
+static void scst_dlm_pr_reg_send_rem_ua(struct scst_device *dev, dlm_lockspace_t *ls,
+					struct scst_dev_registrant *reg, struct scst_dlm_rem_ua *ua)
+{
+	char reg_ua_name[32];
+	struct dlm_ua_lvb *ua_lvb;
+
+	TRACE_ENTRY();
+
+	snprintf(reg_ua_name, sizeof(reg_ua_name), PR_REG_UA_LOCK, reg->dlm_idx,
+		 reg->next_rem_ua_idx);
+
+	/* Always create a new lock. */
+	if (scst_dlm_lock_wait(ls, DLM_LOCK_NL, &ua->lksb, 0, reg_ua_name, NULL) < 0) {
+		PRINT_ERROR("Failed to lock (NL) %s.%s", dev->virt_name, reg_ua_name);
+		return;
+	}
+
+	if (scst_dlm_lock_wait(ls, DLM_LOCK_PW, &ua->lksb,
```

Comment:
Please use "Guard Clause" technique here.

This programming pattern is used to simplify the code by returning early, breaking, or jumping to a label (using goto) when a certain condition is met, rather than nesting the main logic inside an if block or using an else block. This can make the code easier to read and understand, as it reduces the level of indentation and complexity.

So it will look like this:

```
if (scst_dlm_lock_wait(...) < 0) {
        PRINT_ERROR("...");
        goto out;
}

...
```